base/libgit2/strarray.jl

```diff
@@ -1,37 +1,15 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-function StrArrayStruct{T<:AbstractString}(strs::T...)
-    strcount = length(strs)
-    for s in strs
-        if Base.containsnul(s)
-            throw("embedded NULs are not allowed in C strings: $(repr(s))")
-        end
-    end
-    sa_strings = convert(Ptr{Cstring}, Libc.malloc(sizeof(Cstring) * strcount))
-    for i=1:strcount
-        len = length(strs[i])
-        in_ptr  = pointer(String(strs[i]))
-        out_ptr = convert(Ptr{UInt8}, Libc.malloc(sizeof(UInt8) * (len + 1)))
-        unsafe_copy!(out_ptr, in_ptr, len)
-        unsafe_store!(out_ptr, zero(UInt8), len + 1) # NULL byte
-        unsafe_store!(sa_strings, convert(Cstring, out_ptr), i)
-    end
-    return StrArrayStruct(sa_strings, strcount)
+function Base.cconvert{T<:AbstractString}(::Type{Ptr{StrArrayStruct}}, strs::Vector{T})
+    data = [Base.cconvert(Cstring, s) for s in strs]
+    ptrs = [Base.unsafe_convert(Cstring, d) for d in data]
+    sa = Ref(StrArrayStruct(pointer(ptrs), length(ptrs)))
+    sa, data, ptrs
```

Comment:
@tkelman related to your comment below: the idea is that `cconvert` returns a tuple but maintains a reference to the original data.

I agree it's not very pretty: any ideas for a better way to do it?